### ISSUE TITLE
[Bug]: .NET Framework 4.7.2 $Driver not found: ***\node\win32_x64\playwright.cmd

### ISSUE THREAD

**Issue by JohnnyZhang0628** [microsoft/playwright-dotnet#1797]
### Playwright version

1.12

### Operating system

Windows

### What browsers are you seeing the problem on?

Chromium

### What happened? / Describe the bug

Then i run `playwright install`,i got an exception "Microsoft.Playwright assembly was found, but is missing required assets. Please ensure to build your project before running Playwright tool." The same code in .NET 5 app,it's OK.

### Code snippet to reproduce your bug

```shell
using var playwright = await Playwright.CreateAsync();
                await using var browser = await playwright.Chromium.LaunchAsync(new BrowserTypeLaunchOptions
                {
                    Headless = _headless
                });

                // 设置默认窗口大小
                var context = await browser.NewContextAsync(new BrowserNewContextOptions
                {
                    ViewportSize = new ViewportSize() { Width = 1920, Height = 1080 }
                });
```

**Comment by JohnnyZhang0628** [comment]
![image](https://user-images.githubusercontent.com/28375774/139610858-a04c72c6-3221-4597-b19a-07afacd4c061.png)
I try to update the package of "Microsoft.Playwright"  1.16.1 version. It still not work.

**Comment by smitbmx** [comment]
Hi, my question is similar to the current one, could you please explain the correct way of how to set up the custom path for the playwright driver?

Currently the path which works for me is this one:

`C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\13163ddd\37e5e286\assembly\dl3\.playwright`

But how can I add the .playwright correctly under my application, I would like to create something like a portable app that will work out of the box with the Playwright.
I'm able to use the `ExecutablePath` of the `BrowserTypeLaunchOptions` for the bundled browser itself. Can I do something like this for the driver?
I've tried to setup the driver path via Environment, but looks like it doesn't have effect:

```
IEnumerable<KeyValuePair<string, string>> environmentKeyValues = new List<KeyValuePair<string, string>>
                {
                    new ("PLAYWRIGHT_DRIVER_PATH",
                        @"C:\inetpub\wwwroot\My_App\App_Data\tools\playwright\webkit\.playwright\node\win32_x64\playwright.cmd")
                };

                BrowserTypeLaunchOptions browserLaunchOptions = new BrowserTypeLaunchOptions
                {
                    ExecutablePath = @"C:\inetpub\wwwroot\My_App\App_Data\tools\playwright\webkit\Playwright.exe",
                    Env = environmentKeyValues
                };
```
Should the mentioned approach work? If not, how to set up the custom path? Thank you in advance! :)

**Comment by smitbmx** [comment]
> Hi, my question is similar to the current one, could you please explain the correct way of how to set up the custom path for the playwright driver?
> 
> Currently the path which works for me is this one:
> 
> `C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\13163ddd\37e5e286\assembly\dl3\.playwright`
> 
> But how can I add the .playwright correctly under my application, I would like to create something like a portable app that will work out of the box with the Playwright. I'm able to use the `ExecutablePath` of the `BrowserTypeLaunchOptions` for the bundled browser itself. Can I do something like this for the driver? I've tried to setup the driver path via Environment, but looks like it doesn't have effect:
> 
> ```
> IEnumerable<KeyValuePair<string, string>> environmentKeyValues = new List<KeyValuePair<string, string>>
>                 {
>                     new ("PLAYWRIGHT_DRIVER_PATH",
>                         @"C:\inetpub\wwwroot\My_App\App_Data\tools\playwright\webkit\.playwright\node\win32_x64\playwright.cmd")
>                 };
> 
>                 BrowserTypeLaunchOptions browserLaunchOptions = new BrowserTypeLaunchOptions
>                 {
>                     ExecutablePath = @"C:\inetpub\wwwroot\My_App\App_Data\tools\playwright\webkit\Playwright.exe",
>                     Env = environmentKeyValues
>                 };
> ```
> 
> Should the mentioned approach work? If not, how to set up the custom path? Thank you in advance! :)

I'm, also OK to have the `.playwright` under the bin folder, even the value of the next code is the `C:\inetpub\wwwroot\My_App`:
```
DirectoryInfo assemblyDirectory = new(AppContext.BaseDirectory);
string appPath = assemblyDirectory.FullName;
```
Why it's looking under the Temporary ASP.NET Files?

**Comment by smitbmx** [comment]
> > Hi, my question is similar to the current one, could you please explain the correct way of how to set up the custom path for the playwright driver?
> > Currently the path which works for me is this one:
> > `C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\13163ddd\37e5e286\assembly\dl3\.playwright`
> > But how can I add the .playwright correctly under my application, I would like to create something like a portable app that will work out of the box with the Playwright. I'm able to use the `ExecutablePath` of the `BrowserTypeLaunchOptions` for the bundled browser itself. Can I do something like this for the driver? I've tried to setup the driver path via Environment, but looks like it doesn't have effect:
> > ```
> > IEnumerable<KeyValuePair<string, string>> environmentKeyValues = new List<KeyValuePair<string, string>>
> >                 {
> >                     new ("PLAYWRIGHT_DRIVER_PATH",
> >                         @"C:\inetpub\wwwroot\My_App\App_Data\tools\playwright\webkit\.playwright\node\win32_x64\playwright.cmd")
> >                 };
> > 
> >                 BrowserTypeLaunchOptions browserLaunchOptions = new BrowserTypeLaunchOptions
> >                 {
> >                     ExecutablePath = @"C:\inetpub\wwwroot\My_App\App_Data\tools\playwright\webkit\Playwright.exe",
> >                     Env = environmentKeyValues
> >                 };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Should the mentioned approach work? If not, how to set up the custom path? Thank you in advance! :)
> 
> I'm, also OK to have the `.playwright` under the bin folder, even the value of the next code is the `C:\inetpub\wwwroot\My_App`:
> 
> ```
> DirectoryInfo assemblyDirectory = new(AppContext.BaseDirectory);
> string appPath = assemblyDirectory.FullName;
> ```
> 
> Why it's looking under the Temporary ASP.NET Files?

I'm sorry for the many messages, but I've debugged your code, the `Microsoft.Playwright.Helpers.Paths` class and find out that the next code:

```
DirectoryInfo assemblyDirectory = new(AppContext.BaseDirectory);
            if (!assemblyDirectory.Exists || !File.Exists(Path.Combine(assemblyDirectory.FullName, "Microsoft.Playwright.dll")))
            {
                var assemblyLocation = typeof(Playwright).Assembly.Location;
                assemblyDirectory = new FileInfo(assemblyLocation).Directory;
            }
```
returns the next path for the `assemblyDirectory`: `C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\13163ddd\37e5e286\assembly\dl3\215d64e4\006e4772_9bccd701` because the Microsoft.Playwright.dll is there. Is it because I installed the Playwright CLI globally?

**Comment by smitbmx** [comment]
> > Hi, my question is similar to the current one, could you please explain the correct way of how to set up the custom path for the playwright driver?
> > Currently the path which works for me is this one:
> > `C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\13163ddd\37e5e286\assembly\dl3\.playwright`
> > But how can I add the .playwright correctly under my application, I would like to create something like a portable app that will work out of the box with the Playwright. I'm able to use the `ExecutablePath` of the `BrowserTypeLaunchOptions` for the bundled browser itself. Can I do something like this for the driver? I've tried to setup the driver path via Environment, but looks like it doesn't have effect:
> > ```
> > IEnumerable<KeyValuePair<string, string>> environmentKeyValues = new List<KeyValuePair<string, string>>
> >                 {
> >                     new ("PLAYWRIGHT_DRIVER_PATH",
> >                         @"C:\inetpub\wwwroot\My_App\App_Data\tools\playwright\webkit\.playwright\node\win32_x64\playwright.cmd")
> >                 };
> > 
> >                 BrowserTypeLaunchOptions browserLaunchOptions = new BrowserTypeLaunchOptions
> >                 {
> >                     ExecutablePath = @"C:\inetpub\wwwroot\My_App\App_Data\tools\playwright\webkit\Playwright.exe",
> >                     Env = environmentKeyValues
> >                 };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Should the mentioned approach work? If not, how to set up the custom path? Thank you in advance! :)
> 
> I'm, also OK to have the `.playwright` under the bin folder, even the value of the next code is the `C:\inetpub\wwwroot\My_App`:
> 
> ```
> DirectoryInfo assemblyDirectory = new(AppContext.BaseDirectory);
> string appPath = assemblyDirectory.FullName;
> ```
> 
> Why it's looking under the Temporary ASP.NET Files?

I've debugged the [Microsoft.Playwright.Helpers.Paths](https://github.com/microsoft/playwright-dotnet/blob/main/src/Playwright/Helpers/Paths.cs) class and I understood why it works like this.
1) The `AppContext.BaseDirectory` (LIne 36) is correct, it points to the my application folder as `C:\inetpub\wwwroot\My_App`.
2) But the Microsoft.Playwright.dll is not there, it's under the bin folder: `C:\inetpub\wwwroot\My_App\bin`, hence the if-statement (Line 37) cannot find the file there and trying to find the Playwright type location via `typeof(Playwright).Assembly.Location` which is points to the `C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\13163ddd\37e5e286\assembly\dl3\215d64e4\006e4772_9bccd701\Microsoft.Playwright.dll`.
3) Hence the rest of the code expects for the .playwright driver under the Temporary ASP.NET files.

**Comment by vandre** [comment]
I can replicate this issue as well on a Console Application with the latest Nuget Package (Version 1.17.0) and a  .Net Framework Console App targeting .NetFramework 4.8. 
It seems like the install tool is not smart enough to recognize that this is not a .Net Core/5/6 project. 

As a workaround what I did is copy the the .playwright folder to the bin/Debug folder from a .Net Core Project.

**Comment by vandre** [comment]
I figured out what the issue is. It has to do with how packages are handled in Full .Net Framework Projects (4.7, 4.8, etc..)

 By default PackageReference is used for .Net Core/5/6 projects.  .NET Framework projects _support_ PackageReference, but currently [**default** ](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files)to packages.config. 

When the Nuget package for Playwright is installed in a Full .Net Project it creates a packages.config file and the following reference in the .CSPROJ file:
```
    <Reference Include="Microsoft.Playwright, Version=1.17.0.0, Culture=neutral, PublicKeyToken=db12f80f85d8ba23, processorArchitecture=MSIL">
      <HintPath>packages\Microsoft.Playwright.1.17.0\lib\netstandard2.0\Microsoft.Playwright.dll</HintPath>
    </Reference>
```

Compare this to .Net Core/5/6 projects, which default to PackageReference:
```
  <ItemGroup>
    <PackageReference Include="Microsoft.Playwright" Version="1.17.0" />
  </ItemGroup>


```
This causes two separate issues with Playwright on Full .Net Projects:

Issue 1: Playwright install fails with the error described above
Issue 2:  `dotnet build ` does not generate the .playwright folder

The workaround is to [migrate ](https://docs.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference)your full framework project to package reference

Once I migrated my project to PackageReference the CLI tools (`dotnet build`, `playwright install`) worked as expected.
